### PR TITLE
runtime: Executor usage counts retain only single-epoch memory

### PR DESCRIPTION
**Pull request by jon-chuang**
#### Problem
Replaces https://github.com/solana-labs/solana/pull/17044

#### Summary of Changes
~~We evict old executors whenever the cache is `cloned_with_epoch` to prevent old programs or program versions from clogging up the cache potentially for all time.~~

We only retain usage counts from the previous epoch, and discard any previous usage counts.

TODO:
- add tests to ensure old unused are indeed evicted for various `Bank::_new_from_parent`

### CODE REVIEW

**Comment by jackcmay** [comment]
Thanks @jon-chuang , this brings us closer.  To answer your question in the last PR, the executor holds the compiled program, which means each time the program is invoked by an instruction the program is pulled from cache instead of recreating the executor from the program account.  Recreating the program includes re-jitting the program.

I'm wondering if this approach still doesn't address the issue where a very popular program racks up a large use count but then is hardly used.  In this case it would still be used occasionally so wouldn't fall off due to the no-used-for-x-epochs eviction but hangs around because its accumulated count is high but its recent use is very low.  What if we combine the approaches?  What do you think of an eviction strategy where there are two counts maintained, a historical count that comes with the clone at epoch and a current epoch count.  At x epoch boundary the current replaces the historical.  This would allow for lfu programs to continue to persist but large counts would drop off and be replaced with their most recent use counts.  If the new counts warrant keeping them in cache so be it, but large historical counts would be ignored if recent use was drastically different.

**Comment by jackcmay** [comment]
Every instruction invokes its own program, which means each instruction could pull a different program from the cache.

I don't think clone would ever occur at a small epoch because the bank marches on.

**Comment by jon-chuang** [comment]
Another minor issue - possibly we want to set shared to false when we do clone with epoch?

**Comment by jackcmay** [comment]
Maybe, what is your reasoning?

**Comment by jon-chuang** [comment]
Again, the CI is not behaving as expected. Locally:
```
running 1 test
[2021-05-12T08:11:53.239012803Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:54.097594830Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:54.736660114Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:54.771533474Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:55.281720339Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:55.594911485Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:56.223396230Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:56.449344294Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:57.048393371Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:57.453280785Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:58.514813749Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:58.540616990Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:59.001288903Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:59.418939670Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:11:59.833806670Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:00.230775506Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:02.777036051Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:03.193108919Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:03.634258385Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:04.140688573Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:04.546452944Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:05.223744934Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:05.634066496Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:05.981071714Z WARN  solana_core::replay_stage] Vote account 2JKGSfsWhxbwxSj5dXaUN69XxccA8CqDEvnsbt3gJovP does not exist.  Unable to vote
[2021-05-12T08:12:08.357622805Z INFO  local_cluster] truncate validator C's ledger
[2021-05-12T08:12:08.420360293Z INFO  local_cluster] truncate validator A's ledger
[2021-05-12T08:12:08.507535186Z INFO  local_cluster] Removing tower!
[2021-05-12T08:12:08.514780396Z INFO  local_cluster] Restart validator C again!!!
[2021-05-12T08:12:09.262604625Z ERROR solana_core::validator] Rebuilding a new tower from the latest vote account due to failed tower restore: IO Error: No such file or directory (os error 2)
[2021-05-12T08:12:11.968728315Z INFO  local_cluster] collected validator C's votes: {29, 30, 31, 32}
[2021-05-12T08:12:11.968817433Z INFO  local_cluster] Restart validator A again!!!
[2021-05-12T08:12:12.727628266Z ERROR solana_core::validator] Rebuilding a new tower from the latest vote account due to failed tower restore: IO Error: No such file or directory (os error 2)
[2021-05-12T08:12:14.476395967Z INFO  local_cluster] Observed A's votes on: [26, 26, 26, 26, 26, 26, 26, 26, 29, 30]
[2021-05-12T08:12:14.476465670Z INFO  local_cluster] THIS TEST expected violations. And indeed, there was some, because of removed persisted tower.
test test_optimistic_confirmation_violation_without_tower ... ok
```

**Comment by jon-chuang** [comment]
@jackcmay so CI runs perfect now. A bit worrying that twice now CI fails when local tests pass. Is CI a bit wonky at times?

**Comment by jackcmay** [comment]
If you think we should always copy on epoch, why the arc?

**Comment by jon-chuang** [comment]
I don't know in what other scenarios the cache may be copied and modified, for instance, a within-epoch copy would result in multiple references to the same `CowCachedExecutors`.

**Comment by jackcmay** [comment]
It more depends on how often the cached module divert between banks

**Comment by jon-chuang** [comment]
Sorry, my bad, so I think if the epoch doesn't change, we fallback on `clone`, as seen here:
```
impl CowCachedExecutors {
  fn clone_with_epoch(&self, epoch: u64) -> Self {
        let executors_raw = self.read().unwrap();
        if executors_raw.current_epoch() == epoch {
            self.clone()
        } else {
            Self {
                shared: false,
                executors: Arc::new(RwLock::new(executors_raw.clone_with_epoch(epoch))),
            }
        }
    }
```
This returns the old `Arc` that is shared.

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -256,16 +267,20 @@ pub struct Builtins {
...
 
-/// LFU Cache of executors
+/// LFU Cache of executors with single-epoch memory of usage counts
 #[derive(Debug)]
 struct CachedExecutors {
     max: usize,
-    executors: HashMap<Pubkey, (AtomicU64, Arc<dyn Executor>)>,
+    current_epoch: Epoch,
+    // the data stored in the tuple are:
+    // (prev epoch count, current epoch count, Executor)
```
Comment (jackcmay) [inline runtime/src/bank.rs:276]: nit: This this should not just be a struct?

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -283,28 +298,50 @@ impl AbiExample for CachedExecutors {
...
+
+    fn clone_with_epoch(&self, epoch: Epoch) -> Self {
         let mut executors = HashMap::new();
-        for (key, (count, executor)) in self.executors.iter() {
-            executors.insert(
-                *key,
-                (AtomicU64::new(count.load(Relaxed)), executor.clone()),
-            );
+        for (key, (last_epoch_count, count, executor)) in self.executors.iter() {
+            // If the epoch has changed, we store the old count and reset the counts
```
Comment (jackcmay) [inline runtime/src/bank.rs:312]: Can you add a more descriptive description of the new cache strategy here for future readers?

**Review by jackcmay — DISMISSED** [review]
Looks great, just a couple of small nits

**Comment by jon-chuang** [comment]
This is why CI is failing:
```

Lighter validator's latest vote is for slot 5566
--
&nbsp; | thread 'main' panicked at 'assertion failed: `(left == right)`
&nbsp; | left: `5577`,
&nbsp; | right: `5566`', local-cluster/tests/local_cluster.rs:756:13
&nbsp; | stack backtrace:
&nbsp; | 0: rust_begin_unwind
&nbsp; | at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
&nbsp; | 1: core::panicking::panic_fmt
&nbsp; | at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
&nbsp; | 2: core::panicking::assert_failed::inner
&nbsp; | 3: core::panicking::assert_failed
&nbsp; | 4: local_cluster::run_cluster_partition
&nbsp; | 5: core::ops::function::FnOnce::call_once
&nbsp; | 6: serial_test::serial_core
&nbsp; | 7: core::ops::function::FnOnce::call_once
&nbsp; | at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/ops/function.rs:227:5
&nbsp; | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
&nbsp; | FAILED


```

**Comment by jon-chuang** [comment]
@jackcmay Here are my local results for the CI tests:
`test result: ok. 36 passed; 0 failed; 4 ignored; 0 measured; 0 filtered out; finished in 1887.05s`

A bit fed up with CI. Could we just merge?

**Comment by jackcmay** [comment]
@mvines Wanted to get a double-check, this changes the cache behavior but I don't think it is a behavior change from a consensus perspective so no featurization.  The only thing I can think if is some validators take longer than others due to differences in caching and that having an effect on the number of txs they process in a particular slot?  Thoughts?

**Comment by mvines** [comment]
Yep I think this can land ungated.  Please do squash all the commits in here on merge

**Pull request merged**